refactor(compiler): replace most usages of getMutableClone

crisbeto opened the pull request:
Replaces (almost) all of the usages of the deprecated `getMutableClone` function from TypeScript which has started to log deprecation warnings in version 4.8 and will likely be removed in version 5.0. The one place we have left is in the default import handling of ngtsc which will be more difficult to remove.

crisbeto commented:
[Passing TGP](http://test/OCL:468139099:BASE:468161667:1660735002714:8bbf06b1)

devversion reviewed (approved):
LGTM for the trivial tslint, but I also looked at the downlevel transform I added in the past..

crisbeto committed:
refactor(compiler): replace most usages of getMutableClone

Replaces (almost) all of the usages of the deprecated `getMutableClone` function from TypeScript which has started to log deprecation warnings in version 4.8 and will likely be removed in version 5.0. The one place we have left is in the default import handling of ngtsc which will be more difficult to remove.

JoostK commented on packages/compiler-cli/src/transformers/downlevel_decorators_transform/downlevel_decorators_transform.ts:
```diff
@@ -602,3 +589,26 @@ export function getDownlevelDecoratorsTransform(
+  } else if (ts.isSetAccessor(node)) {
+    clone = createSetAccessorDeclaration(modifiers, node.name, node.parameters, node.body);
+  } else {
+    throw new Error(`Unsupported decorated member with kind ${node.kind}`);
```
There's quite a few more `ts.ClassElement` types than covered here. I'm slightly worried that this may result in hard compiler crashes for unexpected code, e.g. currently this fails for decorated getters.

The caller of this function is able to populate a `diagnostics` array, which is used for e.g. computed property names. I'd prefer if we do something similar here to avoid hard crashes.

crisbeto replied:
We don't need to be exhaustive here with all of the `ClassElement` types. It only needs to cover methods, properties, getters and setters. I left the error here so that we get a better error if something slips through, but that's unlikely given that `transformClassDeclaration` only calls  `transformClassElement` for these four node types.


> e.g. currently this fails for decorated getters.

This should be covered by the `} else if (ts.isGetAccessor(node)) {` on line 604.

JoostK replied:
> This should be covered by the } else if (ts.isGetAccessor(node)) { on line 604.

Oh indeed; sorry I missed that one.

And now I notice that the caller is indeed making sure that only supported types are provided, so this should indeed be fine altogether.

JoostK commented on packages/compiler-cli/src/ngtsc/imports/src/default.ts:
```diff
@@ -138,6 +138,7 @@ export class DefaultImportTracker {
         //
         // TODO(alxhub): discuss with the TypeScript team and determine if there's a better way to
         // deal with this issue.
+        // tslint:disable-next-line: ban
```
Do we have follow-up actions planned to remove this usage going forward?

crisbeto replied:
Not yet, but I wanted to all of the actionable ones out of the way so it's easier to deal with this one later.

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_emitter.ts:
```diff
@@ -121,7 +121,22 @@ export class TypeEmitter {
+          } else if (ts.isRegularExpressionLiteral(node)) {
+            clone = ts.factory.createRegularExpressionLiteral(node.text);
+          } else {
+            throw new Error(`Unsupported literal kind ${node.kind}`);
```
I'm not sure that the error message we get with it is much better than the runtime error:

```
Argument of type 'LiteralExpression' is not assignable to parameter of type 'never'.ts(2345)
```

Anyway, I don't think that we need to cover every possible type of literal here.

JoostK replied:
I'd argue that a compile time error is always preferred over a runtime error, since the runtime error stays undiscovered until someone actually uses an unsupported literal expression here.

> Anyway, I don't think that we need to cover every possible type of literal here.

This would mean that this operation becomes fallible, whereas a positive response from `canEmitType` should always let `TypeEmitter.emtType` succeed, otherwise the compiler would crash instead of using a safe fallback approach (we can't fallback if a crash is detected, as the emit strategy to use must be known earlier than the actual emit)

crisbeto replied:
I'd be fine with the compile-time check if it gave us a proper error. I think that the approach above doesn't work, because all of the individual literal node types extend `LiteralExpression`, instead of declaring it as something like `type LiteralExpression = StringLiteral|NumbericLiteral|BigIntLiteral...`.

JoostK replied:
Oh yeah, if it's not a type union then this won't work :-(

crisbeto commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -1317,10 +1317,19 @@ class Scope {
+        throw new Error(`Could not resolve ${node} to an Identifier or a NonNullExpression`);
+      }
+
+      ts.setOriginalNode(clone, res);
+      (clone as any).parent = clone.parent;
```
I think that pulling it out into a function will make it easier for us to reach for it in other places. I'd rather keep it inline for now given that it's only in two places.

JoostK replied:
What I'm mostly concerned about here is that the _intent_ of the code is unclear. Pulling it out into a (documented) function would clear this up.

JoostK reviewed (approved):
LGTM with one nit

crisbeto committed:
fixup! refactor(compiler): replace most usages of getMutableClone

alxhub commented:
This PR was merged into the repository by commit ddb95c51844c5d8aee2187712ba948e41baaf0aa.
